miniapps/navier/navier_particles.cpp

```diff
@@ -0,0 +1,457 @@
+   real_t outlet_length = bc.outlet_start.DistanceTo(
+                             bc.outlet_end); // should be == inlet_length
+
+   Vector inlet_tan(2), outlet_tan(2);
+   inlet_tan = bc.inlet_end;
+   inlet_tan -= bc.inlet_start;
+   inlet_tan /= inlet_length;
+   outlet_tan = bc.outlet_end;
+   outlet_tan -= bc.outlet_start;
+   outlet_tan /= outlet_length;
+
+   real_t t1;
+   Vector p_xn(2), p_xnm1(2), x_int(2), p_xc(2), p_x_int_diff(2);
+   for (int i = 0; i < fluid_particles.GetNParticles(); i++)
+   {
+      X().GetValuesRef(i, p_xn);
+      X(1).GetValuesRef(i, p_xnm1);
+
+      // If outlet_start to outlet_end and x_nm1 to x_n intersect,
+      // apply recirculation
+      if (Get2DSegmentIntersection(bc.outlet_start, bc.outlet_end, p_xnm1, p_xn,
+                                   x_int, &t1))
+      {
+         // Compute the corresponding intersection location on inlet
+         p_xc = 0.0;
+         add(bc.inlet_start, t1*inlet_length, inlet_tan, p_xc);
+
+         // Compute the normal distance from p_xn to x_int, and add
+         p_x_int_diff = p_xn;
+         p_x_int_diff -= x_int;
+         real_t normal_dist = abs(p_x_int_diff*outlet_normal);
+         add(p_xc, normal_dist, inlet_normal, p_xc);
+
+         // Compute tangential distance from p_xn to x_int, and add
+         real_t tan_dist = abs(p_x_int_diff*outlet_tan);
+         add(p_xc, tan_dist, inlet_tan, p_xc);
+
+         // Update the position
+         p_xn = p_xc;
+
+         // Set order to 0, to avoid re-computing x history as well
+         Order()[i] = 0;
+      }
+
+   }
+}
+
+void NavierParticles::ApplyBCs()
+{
+   for (BCVariant &bc_v : bcs)
+   {
+      std::visit(
+         [this](auto &bc)
+      {
+         using T = std::decay_t<decltype(bc)>;
+         if constexpr(std::is_same_v<T, ReflectionBC_2D>)
+         {
+            Apply2DReflectionBC(bc);
+         }
+         else if constexpr(std::is_same_v<T, RecirculationBC_2D>)
+         {
+            Apply2DRecirculationBC(bc);
+         }
+      },bc_v);
+
+   }
+}
+
+NavierParticles::NavierParticles(MPI_Comm comm, int num_particles, Mesh &m)
+   : fluid_particles(comm, num_particles, m.SpaceDimension()),
+     inactive_fluid_particles(comm, 0, m.SpaceDimension()),
+     finder(comm)
+{
+
+   for (int o = 0; o < 3; o++)
+   {
+      beta_k[o].SetSize(4);
+      alpha_k[o].SetSize(3);
+   }
+
+   finder.Setup(m);
+
+   int dim = fluid_particles.GetDim();
+
+   // Initialize kappa, zeta, gamma
+   fp_idx.field.kappa = fluid_particles.AddField(1, Ordering::byVDIM, "kappa");
+   fp_idx.field.zeta = fluid_particles.AddField(1, Ordering::byVDIM, "zeta");
+   fp_idx.field.gamma = fluid_particles.AddField(1, Ordering::byVDIM, "gamma");
+
+   inactive_fluid_particles.AddField(1, Ordering::byVDIM, "kappa");
+   inactive_fluid_particles.AddField(1, Ordering::byVDIM, "zeta");
+   inactive_fluid_particles.AddField(1, Ordering::byVDIM, "gamma");
```

Comment:
Considering how common use case this is of `AddField()`, I would swap the parameters for ordering and name, so the former can be omitted.

Reply:
I see your point, but it could be argued that the ordering is more important than the name (which we consider optional).